Cython/Utility/CommonStructures.c

```diff
@@ -7,98 +7,90 @@ static PyObject* __Pyx_FetchCommonTypeFromSpec(PyType_Spec *spec, PyObject *base
 
 /////////////// FetchCommonType ///////////////
 
-static PyTypeObject* __Pyx_FetchCommonType(PyTypeObject* type) {
-    PyObject* fake_module;
-    PyTypeObject* cached_type = NULL;
-
-    fake_module = PyImport_AddModule((char*) "_cython_" CYTHON_ABI);
+static PyObject *fetch_fake_module(void) {
+    PyObject *fake_module = PyImport_AddModule((char*) "_cython_" CYTHON_ABI);
     if (!fake_module) return NULL;
     Py_INCREF(fake_module);
+    return fake_module;
+}
+
+static void verify_cached_type(PyObject *cached_type,
+                               const char *name,
+                               Py_ssize_t basicsize,
+                               Py_ssize_t expected_basicsize) {
+    if (!PyType_Check(cached_type)) {
+        PyErr_Format(PyExc_TypeError,
+            "Shared Cython type %.200s is not a type object", name);
+    }
+    if (basicsize != expected_basicsize) {
+        PyErr_Format(PyExc_TypeError,
+            "Shared Cython type %.200s has the wrong size, try recompiling",
+            name);
+    }
+}
+
+static PyTypeObject* __Pyx_FetchCommonType(PyTypeObject* type) {
+    PyObject* fake_module;
+    PyTypeObject *cached_type = NULL;
 
+    fake_module = fetch_fake_module();
+    if (!fake_module) goto done;
     cached_type = (PyTypeObject*) PyObject_GetAttrString(fake_module, type->tp_name);
     if (cached_type) {
-        if (!PyType_Check((PyObject*)cached_type)) {
-            PyErr_Format(PyExc_TypeError,
-                "Shared Cython type %.200s is not a type object",
-                type->tp_name);
-            goto bad;
-        }
-        if (cached_type->tp_basicsize != type->tp_basicsize) {
-            PyErr_Format(PyExc_TypeError,
-                "Shared Cython type %.200s has the wrong size, try recompiling",
-                type->tp_name);
-            goto bad;
-        }
-    } else {
-        if (!PyErr_ExceptionMatches(PyExc_AttributeError)) goto bad;
-        PyErr_Clear();
-        if (PyType_Ready(type) < 0) goto bad;
-        if (PyObject_SetAttrString(fake_module, type->tp_name, (PyObject*) type) < 0)
-            goto bad;
-        Py_INCREF(type);
-        cached_type = type;
+        verify_cached_type(
+            (PyObject *)cached_type,
+            type->tp_name,
+            cached_type->tp_basicsize,
+            type->tp_basicsize);
+        goto done;
     }
 
+    if (!PyErr_ExceptionMatches(PyExc_AttributeError)) goto done;
+    PyErr_Clear();
+    if (PyType_Ready(type) < 0) goto done;
+    if (PyObject_SetAttrString(fake_module, type->tp_name, (PyObject *)type) < 0)
+        goto done;
+    Py_INCREF(type);
+    cached_type = type;
+
 done:
-    Py_DECREF(fake_module);
     // NOTE: always returns owned reference, or NULL on error
-    return cached_type;
-
-bad:
+    Py_XDECREF(fake_module);
     Py_XDECREF(cached_type);
-    cached_type = NULL;
-    goto done;
+    return cached_type;
```

Comment:
`XDECREF(x); return x;` ? That combination should never occur in any (readable) code.

Reply:
Maybe when you are returning a borrowed reference and the error case is `NULL`? :) But yeah, in this case this is incorrect. Fixing it!

Reply:
Done